---
title: "TGP Tour"
author: "Alexandra Tidd"
date: "October 28, 2021"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
```

## Interactive development

To simulate installing and loading the package during interactive development...
```{r load_all}
setwd("/working/lab_jonathb/alexandT/tgp/")
devtools::load_all()
```

This package was written using package development conventions from <https://r-pkgs.org/>. 
The functions are all documented, so once the package is loaded you can access the help pages for individual functions, which explain all the arguments. `predict_target_genes()` is the user-facing master function of this package. All other functions are helper functions called by `predict_target_genes()`.
```{r function help}
?predict_target_genes()
```

If you also want the external reference data within the package directory, you can copy it over to mirror my structure...
```{bash copy reference directories}
MyPackageDir=/working/lab_jonathb/alexandT/tgp/
YourPackageDir=path/to/your/copy/of/tgp/
mkdir -p $YourPackageDir/{reference_data,example_data}/data
cp -r $MyPackageDir/reference_data/data/* $YourPackageDir/reference_data/data/
cp -r $MyPackageDir/example_data/data/* $YourPackageDir/example_data/data/
```

To run `predict_target_genes()`...
```{r run master function}
# These paths work if you copied across my structure, as above. The default paths are full paths to my files, so should work the same.
MA <- predict_target_genes(
  trait = "BC",
  outDir = "out/",
  variantsFile = "example_data/data/BC/VariantList.bed",
  driversFile = "example_data/data/BC/Drivers.txt",
  do_scoring = T,
  do_performance = T,
  do_XGBoost = T
  )
```

If you wish to run the internal code of the `R/predict_target_genes.R` script step-by-step for development and debugging, you can add its default arguments to your global environment...
```{r load default arguments}
setwd("/working/lab_jonathb/alexandT/tgp/") ; library(devtools) ; load_all() 
setwd("/working/lab_jonathb/alexandT/tgp") ; 
celltype_of_interest = NULL 
tissue_of_interest = NULL 
trait="BC" 
outDir = "out/" 
variantsFile="/working/lab_jonathb/alexandT/tgp/example_data/data/BC/VariantList.bed" 
driversFile = "/working/lab_jonathb/alexandT/tgp/example_data/data/BC/Drivers.txt" 
referenceDir = "/working/lab_jonathb/alexandT/tgp/reference_data/data/" 
variant_to_gene_max_distance = 2e6 
min_proportion_of_variants_in_top_H3K27ac = 0.05 
do_all_celltypes = F 
do_all_celltypes_in_enriched_tissue = T 
do_scoring = T 
do_performance = T 
do_XGBoost = T 
do_timestamp = F  
```
All functions in the package, unless generic, use the same object names as `predict_target_genes()`, so you can also run the internal code of the helper functions directly as long as you have already run the internal code of `predict_target_genes()` up to the point at which that helper function is called. This allows you to run the complete code line-by-line for debugging package functions.

If you are calling `predict_target_genes()` repeatedly in the same session, you can load the large reference objects `H3K27ac` and `HiChIP` into the global environment once, and then pass them to the function pre-loaded. This prevents redundant re-loading with each call to `predict_target_genes()`.
```{r pre-load reference objects}
# 1. load referenceDir objects
referenceDir = "/working/lab_jonathb/alexandT/tgp/reference_data/data/"
HiChIP <- readRDS(paste0(referenceDir, "HiChIP.rda"))
H3K27ac <- readRDS(paste0(referenceDir, "H3K27ac.rda"))
# 2. pass objects to predict_target_genes for quicker runtime
MA <- predict_target_genes(
  trait = "BC",
  outDir = "out/",
  variantsFile = "example_data/data/BC/VariantList.bed",
  driversFile = "example_data/data/BC/Drivers.txt",
  do_scoring = T,
  do_performance = T,
  do_XGBoost = T,
  HiChIP = HiChIP,
  H3K27ac = H3K27ac
  )
```

## Package data
This package will use both reference genomic annotation datasets and user-provided trait-specific datasets.

### Reference data
#### Internal reference data
Smaller generic reference datasets, including chromosome sizes, GENCODE annotations and REVEL annotations (`ChrSizes`, `TSSs`, `exons`, `introns`, `promoters`, `missense`, `nonsense`, `splicesite`) are stored internally as parsed objects in `R/sysdata.R`. They are accessible when the package is loaded, but not visible due to lazy loading. The reproducible code to generate these objects is in `data-raw/sysdata.R`.

#### External reference data
Larger cell-type-specific reference datasets are stored as local files in  `reference_data/data/`. These are too large to upload to GitHub and would make the package too bulky, so they will be published in a directory to be downloaded alongside the package. The reproducible code to generate these files is in `reference_data/data-raw/`
```{r local reference directory, eval = T}
list.files("/working/lab_jonathb/alexandT/tgp/reference_data/data/")
```

### User-provided data
There is one required user-provided file for the `predict_target_genes()` function: a list of trait variants (the `variantsFile` argument). A list of trait drivers (the `driversFile` argument) is needed only if `do_performance = T` is set. Example inputs can be found in my local `example_data/data/` directory. Full paths to these are set as the default arguments of the function.

#### Trait variants 
The variants file should be a BED file with metadata columns for the variant name and the credible set to which it belongs.
```{bash variantsFile structure, eval = T}
head /working/lab_jonathb/alexandT/tgp/example_data/data/BC/VariantList.bed
```

#### Trait drivers
The drivers file should be a file with a single column of driver gene symbols. These symbols should be GENCODE-compatible.
```{bash driversFile structure, eval = T}
head /working/lab_jonathb/alexandT/tgp/example_data/data/BC/Drivers.txt
```
